### PR TITLE
YAML: Convert ip_allow.config

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This is a first pass at converting `ip_allow.config` to YAML. The early commits will need to be pulled out in to separate PRs once this is in final form.

This allows configuration in the old format or YAML. It guesses the contents based on the first line. If it's "#" or "# ats" then it uses the old format, otherwise it uses YAML.

This still needs documentation updates.

This ended up require the break out of other PRs. The current outstanding one is #4206.
Merged: #4175, #4176, #4149

### CODE REVIEW

File: proxy/IPAllow.cc [hunk 1]
```diff
@@ -42,8 +44,64 @@ SignalError(ts::BufferWriter &w, bool &flag)
...
+  return w.print("{}[{}]", obj->MODULE_NAME, obj->get_config_file());
+}
+
+BufferWriter &
+bwformat(BufferWriter &w, BWFSpec const &spec, YAML::Mark const &mark)
+{
+  return w.print("Line {}", mark.line);
+}
+
+BufferWriter &
```
Comment (SolidWallOfCode) [inline proxy/IPAllow.cc]: This should probably be moved to `bw_std_format.h` in a separate PR as it's not specific to this PR.

File: proxy/IPAllow.cc [hunk 2]
```diff
@@ -42,8 +44,64 @@ SignalError(ts::BufferWriter &w, bool &flag)
...
+bwformat(BufferWriter &w, BWFSpec const &spec, std::error_code const &ec)
+{
+  return w.print("[{}:{}]", ec.value(), ec.message());
+}
+
+} // namespace ts
+
+namespace YAML
```
Comment (SolidWallOfCode) [inline proxy/IPAllow.cc]: Need to comment the code, but this is needed in order to pass a `TextView` to the array operator on a node.

File: proxy/IPAllow.cc [hunk 3]
```diff
@@ -195,24 +253,203 @@ IpAllow::Print()
...
-  IpAddr addr2;
-  bool alarmAlready = false;
-  ts::LocalBufferWriter<1024> bw_err;
-
   // Table should be empty
   ink_assert(_src_map.count() == 0 && _dst_map.count() == 0);
 
-  file_buff = readIntoBuffer(config_file_path, "ip-allow", &file_size);
+  std::error_code ec;
+  ts::file::path fpath{config_file_path};
```
Comment (SolidWallOfCode) [inline proxy/IPAllow.cc]: Change `config_file_path` from `ats_scoped_str` to `ts::file::path` and skip this step?

File: lib/ts/ts_file.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/** @file
```
[the remark this answers is not in the document]
Comment (SolidWallOfCode) [reply lib/ts/ts_file.cc]: `std::filesystem` is still experimental and I've already found one major difference between the experimental version and the standard version. 

I don't think we have any of these in `libtsutil` currently.

**Comment by SolidWallOfCode** [comment]
Added documentation, updated the default configuration.

**Comment by SolidWallOfCode** [comment]
I am unable to find anyway to link with `std::experimental::filesystem` for the FreeBSD builds. I've tried various approaches, nothing, `c++`, `stdc++fs`, `c++fs`, `c++experimental`, which were recommend in various places, but none were successful. If anyone else wants to try, tweak the "AC_SUBST(STD_FS_LIB, [stdc++fs])" line in `configure.ac`.

**Comment by zwoop** [comment]
This was a great exercise, because it found a fundamental problem with FreeBSD 11.1. Ask and I believe that upgrading to 11.2 can fix this, and I’m working on that. For ATS v8.0and later, we will have to require fbsd 11.2 or later.

**Comment by SolidWallOfCode** [comment]
This has been waiting on other PRs that spun out of it, I think it's down to just #4206, after which I'll rebase and do a full submission.

**Comment by SolidWallOfCode** [comment]
Rebased and ready for review.

**Comment by randall** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
Hmmm, this failed clang-format, but when I tried to update it, `make clang-format` made no changes. We'll see if it fails again.

**Comment by SolidWallOfCode** [comment]
[ci approve autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
[approve ci]

**Comment by SolidWallOfCode** [comment]
[approve ci rat]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

File: configs/ip_allow.schema.json [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      "items": {
+        "$ref": "#/definitions/rule"
+      }
+    }
+  },
+  "required": [ "ip_addr_acl" ],
+  "definitions": {
+    "range": {
+      "description": "A range of IP addresses in a single family.",
+  "type": "string"
```
Comment (maskit) [inline configs/ip_allow.schema.json]: Indent level seems wrong.

**Comment by randall** [comment]
@SolidWallOfCode Can this be rebased?

**Comment by SolidWallOfCode** [comment]
[approve ci clang-analyzer]

**Comment by SolidWallOfCode** [comment]
[approve ci clang-analyzer]

File: proxy/IPAllow.cc [hunk 6]
```diff
@@ -42,8 +45,64 @@ SignalError(ts::BufferWriter &w, bool &flag)
...
+bwformat(BufferWriter &w, BWFSpec const &spec, std::error_code const &ec)
+{
+  return w.print("[{}:{}]", ec.value(), ec.message());
+}
+
+} // namespace ts
+
+namespace YAML
+{
+template <> struct convert<ts::TextView> {
```
Comment (zwoop) [inline proxy/IPAllow.cc]: Does this need to be TextView ? Can it not be std::string_view?

**Pull request merged**